server/log_operation_manager.go

```diff
@@ -92,11 +215,60 @@ func (l LogOperationManager) getLogIDs(ctx context.Context) ([]int64, error) {
 	return logIDs, nil
 }
 
-func (l LogOperationManager) getLogsAndExecutePass(ctx context.Context) error {
-	logIDs, err := l.getLogIDs(ctx)
+func (l *LogOperationManager) masterFor(ctx context.Context, allIDs []int64) ([]int64, error) {
+	if l.info.Registry.ElectionFactory == nil {
+		return allIDs, nil
+	}
+	if l.tracker == nil {
+		glog.Infof("creating mastership tracker for %v", allIDs)
+		l.tracker = util.NewMasterTracker(allIDs)
+	}
+
+	// Synchronize the set of configured log IDs with those we are tracking mastership for.
+	for _, logID := range allIDs {
+		if l.electionRunner[logID] != nil {
+			continue
+		}
+		glog.Infof("create master election goroutine for %v", logID)
+		innerCtx, cancel := context.WithCancel(ctx)
+		election, err := l.info.Registry.ElectionFactory.NewElection(innerCtx, logID)
+		if err != nil {
+			cancel()
+			return nil, fmt.Errorf("failed to create election for %d: %v", logID, err)
+		}
+		l.electionRunner[logID] = &electionRunner{
+			logID:    logID,
+			info:     &l.info,
+			tracker:  l.tracker,
+			cancel:   cancel,
+			wg:       &l.runnerWG,
+			election: election,
+		}
+		l.runnerWG.Add(1)
+		go l.electionRunner[logID].Run(innerCtx)
+	}
+
+	held := l.tracker.Held()
+	glog.V(1).Infof("acting as master for %d / %d: %s", len(held), len(allIDs), l.tracker)
+	return held, nil
+}
+
+func (l *LogOperationManager) getLogsAndExecutePass(ctx context.Context) error {
+	allIDs, err := l.getLogIDs(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve full list of log IDs: %v", err)
+	}
+	logIDs, err := l.masterFor(ctx, allIDs)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to determine log IDs we're master for: %v", err)
 	}
+	l.heldMutex.Lock()
```

Comment:
could move this block out into an `updateMasterIDs(logIDs)` (or something func), that'd trim this func size down and let you do the standard immediate `defer` thing to release the mutex lock.